streamrouter/router.py

```diff
@@ -105,306 +155,564 @@ def snapshot_url(self):
+class NativeRoute(NativeObject):
+    """
+    A wrapper around a native route object.
+    """
 
-    possible_streams = {
-        Route.STREAM_FORMAT_HLS: 'playlist.m3u8',
-        Route.STREAM_FORMAT_MP4: 'stream.mp4',
-        Route.STREAM_FORMAT_MJPEG: 'stream.mjpeg',
-        Route.LIVE_SNAPSHOT: 'snapshot.jpg'
+    stream_format_map = {
+        Route.STREAM_FORMAT_HLS: lib.STREAM_FORMAT_HLS,
+        Route.STREAM_FORMAT_MP4: lib.STREAM_FORMAT_MP4,
+        Route.STREAM_FORMAT_MJPEG: lib.STREAM_FORMAT_MJPEG,
+        Route.LIVE_SNAPSHOT: lib.STREAM_FORMAT_LIVE_SNAPSHOT,
     }
 
-    def __init__(self, resource, token, master, proto='https'):
-        super().__init__(token, resource, proto=proto)
-        self.master = master
+    def __init__(self, raw_ptr, free_raw_ptr=True, proto='https'):
+        if free_raw_ptr:
+            free_func = lib.srl__route__free
+        else:
+            free_func = None
 
-    def _service(self, stream_format):
-        return {'name': 'rtsp-master', 'service': self.master}
+        super().__init__(raw_ptr, free_func=free_func)
 
+        self.proto = proto
 
-class EdgeRoute(RtspconRoute):
-    url_mask = '{proto}://{host}/{master}/{camera_id}/{stream_name}?token={token}'
+    def is_supported_format(self, stream_format):
+        """
+        Check if a given stream format is supported by this type of route.
+        """
+        native_stream_format = self.stream_format_map[stream_format]
 
-    possible_streams = {
-        Route.STREAM_FORMAT_HLS: 'playlist.m3u8',
-        Route.STREAM_FORMAT_MP4: 'stream.mp4',
-    }
+        assert self.raw_ptr is not None
 
-    def __init__(self, resource, token, master, rtsp_edge, mp4_edge, proto='https'):
-        super().__init__(resource, token, master, proto=proto)
-        self.rtsp_edge = rtsp_edge
-        self.mp4_edge = mp4_edge
+        res = lib.srl__route__is_supported_format(self.raw_ptr, native_stream_format)
 
-    def _service(self, stream_format):
-        if self.master is None:
-            raise ServiceUnavailable('rtsp-master')
-        if stream_format == self.STREAM_FORMAT_MP4:
-            return {'name': 'mp4-edge', 'service': self.mp4_edge}
-        return {'name': 'rtsp-edge', 'service': self.rtsp_edge}
+        return bool(res)
 
-    @property
-    def master_key(self):
-        return self.master.id.replace('rtsp-master-', '')
+    def get_url(self, stream_format):
+        """
+        Get stream URL for a given format.
+        """
+        if not self.is_supported_format(stream_format):
+            raise UnsupportedStreamFormat(stream_format)
+
+        native_stream_format = self.stream_format_map[stream_format]
+
+        scheme = self.proto.encode('utf-8')
+
+        return get_string(
+            lib.srl__route__get_url_with_custom_scheme,
+            self.raw_ptr,
+            native_stream_format,
+            scheme)
+
+
+class EdgeRoute(Route, NativeObject):
```

Comment:
same as with services classes - all route classes has almost identical init and del method

Reply:
They look the same but they use different native methods. Anyway, I'll try to refactor them somehow to avoid the code bloat.